model/src/w3srcemd.F90

```diff
@@ -1378,9 +1395,96 @@ SUBROUTINE W3SRCE ( srce_call, IT, JSEA, IX, IY, IMOD,          &
                WRITE(740+IAPROC,*) 'min/max/sum(VDBT)=', minval(VDBT), maxval(VDBT), sum(VDBT)
         END IF
 #endif
+         IF (srce_call .eq. srce_imp_pre) THEN
+#ifdef W3_PDLIB
+      IF (LSLOC) THEN
+        IF (IMEM == 1) THEN
+          SIDT  = PDLIB_SI(JSEA) * DTG 
+          DO IK = 1, NK
+            JAC = CLATSL/CG1(IK) 
+            DO ITH = 1, NTH
+              ISP = ITH + (IK-1)*NTH
+              VD(ISP) = MIN(0., VD(ISP))
+              IF (LNEWLIMITER) THEN
+                MAXDAC = MAX(DAM(ISP),DAM2(ISP))
+              ELSE
+                MAXDAC = DAM(ISP)
+              ENDIF
+              FAKS   = DTG / MAX ( 1. , (1.-DTG*VD(ISP)))
+              DVS    = VS(ISP) * FAKS
+              DVS    = SIGN(MIN(MAXDAC,ABS(DVS)),DVS)
+              PreVS  = DVS / FAKS
+              eVS    = PreVS / CG1(IK) * CLATSL
+              eVD    = MIN(0.,VD(ISP))
+              B_JAC(ISP,JSEA)                   = B_JAC(ISP,JSEA) + SIDT * (eVS - eVD*SPEC(ISP)*JAC)
+              ASPAR_JAC(ISP,PDLIB_I_DIAG(JSEA)) = ASPAR_JAC(ISP,PDLIB_I_DIAG(JSEA)) - SIDT * eVD
+#endif
 
+#ifdef W3_DB1
+                eVS = VSDB(ISP) * JAC
```

Comment:
ISP is not defined unless PDLIB is also used.  Perhaps you can't get into this part of the loop unless you are using PDLIB, but I think this code should look like this where the whole code is in one PDLIB ifdef section:  

```
#ifdef W3_PDLIB
       IF (LSLOC) THEN
         IF (IMEM == 1) THEN
           SIDT  = PDLIB_SI(JSEA) * DTG 
           DO IK = 1, NK
             JAC = CLATSL/CG1(IK) 
             DO ITH = 1, NTH
               ISP = ITH + (IK-1)*NTH
               VD(ISP) = MIN(0., VD(ISP))
               IF (LNEWLIMITER) THEN
                 MAXDAC = MAX(DAM(ISP),DAM2(ISP))
               ELSE
                 MAXDAC = DAM(ISP)
               ENDIF
               FAKS   = DTG / MAX ( 1. , (1.-DTG*VD(ISP)))
               DVS    = VS(ISP) * FAKS
               DVS    = SIGN(MIN(MAXDAC,ABS(DVS)),DVS)
               PreVS  = DVS / FAKS
               eVS    = PreVS / CG1(IK) * CLATSL
               eVD    = MIN(0.,VD(ISP))
               B_JAC(ISP,JSEA)                   = B_JAC(ISP,JSEA) + SIDT * (eVS - eVD*SPEC(ISP)*JAC)
               ASPAR_JAC(ISP,PDLIB_I_DIAG(JSEA)) = ASPAR_JAC(ISP,PDLIB_I_DIAG(JSEA)) - SIDT * eVD

 #ifdef W3_DB1
                 eVS = VSDB(ISP) * JAC
                 eVD = MIN(0.,VDDB(ISP))
 #endif

 #ifdef W3_DB2
                 eVS = VSDB(ISP) * JAC
                 eVD = MIN(0.,VDDB(ISP))
 #endif

               B_JAC(ISP,JSEA)                   = B_JAC(ISP,JSEA) + SIDT * eVS
               ASPAR_JAC(ISP,PDLIB_I_DIAG(JSEA)) = ASPAR_JAC(ISP,PDLIB_I_DIAG(JSEA)) - SIDT * eVD

 #ifdef W3_TR1
                 eVS = VSTR(ISP) * JAC 
                 eVD = VDTR(ISP)
 #endif

               B_JAC(ISP,JSEA)                   = B_JAC(ISP,JSEA) + SIDT * eVS
               ASPAR_JAC(ISP,PDLIB_I_DIAG(JSEA)) = ASPAR_JAC(ISP,PDLIB_I_DIAG(JSEA)) - SIDT * eVD
             END DO
           END DO

         ELSEIF (IMEM == 2) THEN
           SIDT   = PDLIB_SI(JSEA) * DTG
           DO IK=1,NK
             JAC = CLATSL/CG1(IK)
             DO ITH=1,NTH
               ISP=ITH + (IK-1)*NTH
               VD(ISP) = MIN(0., VD(ISP))
               IF (LNEWLIMITER) THEN
                 MAXDAC    = MAX(DAM(ISP),DAM2(ISP))
               ELSE
                 MAXDAC    = DAM(ISP)
               ENDIF
               FAKS      = DTG / MAX ( 1. , (1.-DTG*VD(ISP)))
               DVS       = VS(ISP) * FAKS
               DVS       = SIGN(MIN(MAXDAC,ABS(DVS)),DVS)
               PreVS     = DVS / FAKS
               eVS = PreVS / CG1(IK) * CLATSL
               eVD = VD(ISP)
 #ifdef W3_DB1
                 eVS = eVS + DBLE(VSDB(ISP)) * JAC
                 eVD = evD + MIN(0.,DBLE(VDDB(ISP)))
 #endif
 #ifdef W3_DB2
                 eVS = eVS + DBLE(VSDB(ISP)) * JAC
                 eVD = evD + MIN(0.,DBLE(VDDB(ISP)))
 #endif

               B_JAC(ISP,JSEA)          = B_JAC(ISP,JSEA) + SIDT * (eVS - eVD*VA(ISP,JSEA))
               ASPAR_DIAG_ALL(ISP,JSEA) = ASPAR_DIAG_ALL(ISP,JSEA) - SIDT * eVD
             END DO
           END DO
         ENDIF
       ENDIF 
 #endif
```